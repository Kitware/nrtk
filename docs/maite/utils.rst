=====
Utils
=====

-----------------------------
NRTK perturber CLI Entrypoint
-----------------------------

.. click:: nrtk.interop.maite.utils.bin.nrtk_perturber_cli:nrtk_perturber_cli
    :prog: nrtk-perturber
    :nested: full

------------------------------------
Augmented MAITE dataset(s) generator
------------------------------------

.. automodule:: nrtk.interop.maite.utils.nrtk_perturber
    :members:
    :special-members:
