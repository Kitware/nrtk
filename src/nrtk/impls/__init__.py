"""Module for all implementations of nrtk interfaces"""
