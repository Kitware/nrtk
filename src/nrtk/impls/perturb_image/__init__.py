"""Module for all PerturbImage implementations"""
