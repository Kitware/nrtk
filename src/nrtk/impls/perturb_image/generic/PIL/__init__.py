"""Module for PIL based implementations of PerturbImage"""
