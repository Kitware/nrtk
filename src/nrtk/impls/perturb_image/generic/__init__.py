"""Module for all generic implementations of PerturbImage"""
