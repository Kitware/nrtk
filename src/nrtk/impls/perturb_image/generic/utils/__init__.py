"""Package for nrtk generic perturber utils needed for carrying out pertubation support operations"""
