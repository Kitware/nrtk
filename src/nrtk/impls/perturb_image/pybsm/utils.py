"""Utility functions for PyBSM sensor and scenario configuration management.

This module provides helper functions for loading and managing default PyBSM sensor
and scenario configurations from JSON configuration files.

Functions:
    default_sensor_scenario: Load default sensor and scenario configurations from config files.

Example usage:
    from nrtk.impls.perturb_image.pybsm.utils import default_sensor_scenario

    # Load UAV configuration
    sensor, scenario = default_sensor_scenario("uav")

    # Use in perturbation factory
    perturber_factory = CustomPybsmPerturbImageFactory(
        sensor=sensor,
        scenario=scenario,
        theta_keys=["ground_range"],
        thetas=[ground_range_values]
    )
"""

from __future__ import annotations

import json
from pathlib import Path

from nrtk.impls.perturb_image.pybsm.scenario import PybsmScenario
from nrtk.impls.perturb_image.pybsm.sensor import PybsmSensor


def default_sensor_scenario(config_type: str, data_dir: str | None = None) -> tuple[PybsmSensor, PybsmScenario]:  # noqa C901
    """Load default sensor and scenario configurations for a given configuration type.

    This function loads pre-defined sensor and scenario configurations from JSON files
    located in the data directory. The configuration files should contain 'sensor' and
    'scenario' keys with the appropriate PyBSM configuration parameters.

    Currently supported configuration types:
    - "uav": Unmanned Aerial Vehicle configuration with typical parameters for
             drone-based imaging systems

    Future configuration types may include:
    - "satellite": Satellite-based imaging system configuration
    - "uav_high": High-altitude UAV configuration
    - "ground": Ground-based imaging system configuration

    Args:
        config_type: String identifier for the configuration type to load.
                    Must be one of the supported configuration types.
        data_dir: Optional path to the directory containing configuration files.
                 If None, the function will search in default locations.
                 Can be a string path or pathlib.Path object.

    Returns:
        A tuple containing:
            - PybsmSensor: Configured sensor object with loaded parameters
            - PybsmScenario: Configured scenario object with loaded parameters

    Raises:
        ValueError: If the configuration type is not supported or the
                   configuration file is missing required keys.
        FileNotFoundError: If the configuration file cannot be found.
        json.JSONDecodeError: If the configuration file contains invalid JSON.
        IOError: If there's an error reading the configuration file.
    """
    # Define supported configuration types
    supported_configs = {
        "uav": "uav_default_config_nrtk.json",
        # "satellite": "satellite_default_config_nrtk.json",
        # "uav_high": "uav_high_default_config_nrtk.json",
        # "ground": "ground_default_config_nrtk.json",
    }

    # Validate configuration type
    if config_type not in supported_configs:
        available_configs = ", ".join(supported_configs.keys())
        raise ValueError(
            f"Unsupported configuration type: '{config_type}'. Available configurations: {available_configs}",
        )

    # Determine configuration file path
    config_filename = supported_configs[config_type]

    # If data_dir is provided, use it as the primary location
    if data_dir is not None:
        data_path = Path(data_dir)
        if not data_path.is_dir():
            raise ValueError(f"Specified data directory does not exist: {data_dir}")

        config_path = data_path / config_filename
        if not config_path.exists():
            raise FileNotFoundError(
                f"Configuration file '{config_filename}' not found in specified data directory: {data_dir}",
            )
    else:
        # Try different potential locations for the config file
        potential_paths = [
            Path("./data") / config_filename,  # Current notebook directory
            Path(__file__).parent / "data" / config_filename,  # Relative to this module
        ]

        config_path = None
        for path in potential_paths:
            if path.exists():
                config_path = path
                break

        if config_path is None:
            searched_paths = [str(p) for p in potential_paths]
            raise FileNotFoundError(
                f"Configuration file '{config_filename}' not found. Searched paths: {searched_paths}",
            )

    # Load and parse configuration file
    try:
        with open(config_path, encoding="utf-8") as file:
            config_data = json.load(file)
    except json.JSONDecodeError as e:
        raise json.JSONDecodeError(
            f"Invalid JSON in configuration file '{config_path}': {e}",
            e.doc,
            e.pos,
        ) from e
    except OSError as e:
        raise OSError(
            f"Error reading configuration file '{config_path}': {e}",
        ) from e

    # Validate required keys exist
    required_keys = ["sensor", "scenario"]
    missing_keys = [key for key in required_keys if key not in config_data]
    if missing_keys:
        raise ValueError(
            f"Configuration file '{config_path}' is missing required keys: {missing_keys}",
        )

    # Extract sensor and scenario configurations
    sensor_config = config_data["sensor"]
    scenario_config = config_data["scenario"]

    # Create and return sensor and scenario objects
    try:
        sensor = PybsmSensor.from_config(sensor_config)
        scenario = PybsmScenario.from_config(scenario_config)
    except Exception as e:
        raise RuntimeError(
            f"Error creating sensor or scenario objects from configuration: {e}",
        ) from e
    return sensor, scenario
