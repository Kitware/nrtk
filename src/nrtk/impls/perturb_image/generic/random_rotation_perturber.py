"""Defines RandomRotationPerturber to apply Albumentations' `Rotate` transformation to input images.

Classes:
    RandomRotationPerturber: A perturbation class for applying the `Rotate` transformation from Albumentations.

Dependencies:
    - albumentations: For the underlying perturbations
    - numpy: For numerical operations and random number generation.
    - smqtk_image_io.AxisAlignedBoundingBox: For handling and adjusting bounding boxes.
    - nrtk.interfaces.perturb_image.PerturbImage: Base class for perturbation algorithms.
    - nrtk.impls.perturb_image.generic.albumentations_perturber: Base implementation for Albumentations perturbers.
"""

from __future__ import annotations

__all__ = ["RandomRotationPerturber"]

from collections.abc import Sequence
from typing import Any

import numpy as np
from typing_extensions import override

from nrtk.impls.perturb_image.generic.albumentations_perturber import AlbumentationsPerturber


class RandomRotationPerturber(AlbumentationsPerturber):
    """RandomRotationPerturber applies a Rotation transformation from Albumentations.

    Attributes:
        limit (float | tuple[float, float]):
            Maximum rotation angle in degrees.
        probability (float):
            Probability of applying the rotation transformation.
        seed (int):
            An optional seed for reproducible results.
        fill (numpy.array):
            Background color fill for RGB image.

    Methods:
        perturb:
            Applies the specified to an input image.
        __call__:
            Calls the perturb method with the given input image.
        get_config:
            Returns the current configuration of the RandomRotationPerturber instance.
    """

    def __init__(
        self,
        limit: float | tuple[float, float] = 0.0,
        probability: float = 1.0,
        fill: Sequence[int] | None = None,
        seed: int | None = None,
    ) -> None:
        """RandomRotationPerturber applies a Rotation transformation from Albumentations.

        Args:
            limit (float | tuple[float, float]):
                Maximum rotation angle in degrees. For example, 90.0 value means the image can be rotated
                by any angle between [-90, 90] degrees and (0.0, 90.0) means the image can be rotated
                by any angle between [0, 90] degrees. Default value is 0.0, which means no rotation.
            probability (float):
                Probability of applying the rotation transformation. A value of 1.0 means it will always
                apply the rotation, while a value of 0.0 means it will never apply the rotation. Default
                value is 1.0.
            fill (numpy.array | None):
                Background color fill for RGB image. (0 to 255 for each channel).
                Default value is a black background ([0, 0, 0]).
            seed (int | None):
                An optional seed for reproducible results. Defaults to None, which means the results are
                non-deterministic.

        Raises:
            :raises ValueError: Rotation probability must be between 0.0 and 1.0 inclusive.
            :raises ValueError: Color fill values must be integers between 0 and 255 inclusive.
        """
        if not 0.0 <= probability <= 1.0:
            raise ValueError("Rotation probability must be between 0.0 and 1.0 inclusive.")

        if fill is None:
            fill = [0, 0, 0]
        fill_arr = np.asarray(fill)
        if not ((fill_arr >= 0) & (fill_arr <= 255)).all():
            raise ValueError("Color fill values must be integers between 0 and 255 inclusive.")

        super().__init__(
            perturber="Rotate",
            parameters={"limit": limit, "p": probability, "fill": fill},
            seed=seed,
        )

        self.limit = limit
        self.probability = probability
        self.fill = fill

    @override
    def get_config(self) -> dict[str, Any]:
        """Returns the current configuration of the RandomRotationPerturber instance.

        Returns:
            :return dict[str, Any]: Configuration dictionary with current settings.
        """
        cfg = dict()
        cfg["limit"] = self.limit
        cfg["probability"] = self.probability
        cfg["fill"] = self.fill
        cfg["seed"] = self.seed
        return cfg
