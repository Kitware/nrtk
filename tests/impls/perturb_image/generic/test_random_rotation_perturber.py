import unittest.mock as mock
from collections.abc import Hashable, Sequence
from contextlib import AbstractContextManager
from contextlib import nullcontext as does_not_raise
from typing import Any
from unittest.mock import MagicMock

import numpy as np
import pytest
from PIL import Image
from smqtk_core.configuration import configuration_test_helper
from smqtk_image_io.bbox import AxisAlignedBoundingBox
from syrupy.assertion import SnapshotAssertion

from nrtk.impls.perturb_image.generic.random_rotation_perturber import RandomRotationPerturber
from nrtk.utils._exceptions import AlbumentationsImportError
from tests.impls import INPUT_TANK_IMG_FILE_PATH as INPUT_IMG_FILE_PATH
from tests.impls.perturb_image.test_perturber_utils import perturber_assertions


@pytest.mark.skipif(not RandomRotationPerturber.is_usable(), reason=str(AlbumentationsImportError()))
class TestRandomRotationPerturber:
    @pytest.mark.parametrize(
        ("limit", "probability", "fill", "expectation"),
        [
            (
                90,
                -0.5,
                [255, 0, 0],
                pytest.raises(
                    ValueError,
                    match=r"Rotation probability must be between 0.0 and 1.0 inclusive.",
                ),
            ),
            (
                90,
                1.5,
                [255, 0, 0],
                pytest.raises(
                    ValueError,
                    match=r"Rotation probability must be between 0.0 and 1.0 inclusive.",
                ),
            ),
            (
                0,
                1.0,
                [355, 0, 0],
                pytest.raises(
                    ValueError,
                    match=r"Color fill values must be integers between 0 and 255 inclusive.",
                ),
            ),
            (
                0,
                1.0,
                [-155, 0, 0],
                pytest.raises(
                    ValueError,
                    match=r"Color fill values must be integers between 0 and 255 inclusive.",
                ),
            ),
            (0, 1.0, [255, 0, 0], does_not_raise()),
            (150, 0.0, [255, 0, 0], does_not_raise()),
            (120, 0.5, [255, 0, 0], does_not_raise()),
            (500, 1.0, [255, 0, 0], does_not_raise()),
            ((0, 90), 1.0, [255, 0, 0], does_not_raise()),
            ((-90, 0), 1.0, [255, 0, 0], does_not_raise()),
            ((-180, 270), 1.0, [255, 0, 0], does_not_raise()),
        ],
    )
    def test_perturber_validity(
        self,
        limit: float | tuple[float, float],
        probability: float,
        fill: Sequence[int],
        expectation: AbstractContextManager,
    ) -> None:
        """Raise appropriate errors for invalid perturber arguments."""
        with expectation:
            RandomRotationPerturber(
                limit=limit,
                probability=probability,
                fill=fill,
            )

    def test_bbox_transform(self, snapshot: SnapshotAssertion) -> None:
        label_dict_1: dict[Hashable, float] = {"label": 1.0}
        label_dict_2: dict[Hashable, float] = {"label": 2.0}
        bboxes = [
            AxisAlignedBoundingBox((1, 1), (2, 3)),
            AxisAlignedBoundingBox((3, 2), (6, 8)),
        ]
        labels = [label_dict_1, label_dict_2]
        image = np.ones((30, 30, 3)).astype(np.uint8)
        inst = RandomRotationPerturber(
            limit=(180, 180),
            probability=1.0,
            seed=12,
        )
        image_out, bboxes_transformed = inst.perturb(
            image=image,
            boxes=zip(bboxes, labels, strict=False),
        )
        _, bboxes_reverted = inst.perturb(image=image_out, boxes=bboxes_transformed)

        if bboxes_reverted:
            bboxes_reverted = list(bboxes_reverted)
            assert len(bboxes_reverted) == len(bboxes)
            for bbox, label, reverted in zip(bboxes, labels, bboxes_reverted, strict=False):
                assert bbox == reverted[0]
                assert label == reverted[1]

        assert bboxes_transformed == snapshot

    def test_identity_operation(self) -> None:
        """Test that the perturber does not change the image when limit is 0."""
        image = Image.open(INPUT_IMG_FILE_PATH)
        image = np.array(image)
        inst = RandomRotationPerturber()
        out_image = perturber_assertions(perturb=inst.perturb, image=image)
        perturber_assertions(perturb=inst, image=image, expected=out_image)

    def test_consistency(self) -> None:
        """Run perturber twice with consistent seed to ensure repeatable results."""
        image = np.ones((3, 3, 3)).astype(np.uint8)

        limit = 300
        probability = 1.0
        seed = 12

        # Test perturb interface directly
        inst = RandomRotationPerturber(
            limit=limit,
            probability=probability,
            seed=seed,
        )
        out_image = perturber_assertions(perturb=inst.perturb, image=image)

        # Test callable
        perturber_assertions(
            perturb=RandomRotationPerturber(
                limit=limit,
                probability=probability,
                seed=seed,
            ),
            image=image,
            expected=out_image,
        )

    def test_regression(self, psnr_tiff_snapshot: SnapshotAssertion) -> None:
        """Regression testing results to detect API changes."""
        grayscale_image = Image.open(INPUT_IMG_FILE_PATH)
        image = Image.new("RGB", grayscale_image.size)
        image.paste(grayscale_image)
        image = np.array(image)
        inst = RandomRotationPerturber(
            limit=90,
            probability=1.0,
            seed=12,
        )
        out_img = perturber_assertions(
            perturb=inst.perturb,
            image=image,
        )
        psnr_tiff_snapshot.assert_match(out_img)

    @pytest.mark.parametrize(
        ("perturber", "limit", "probability", "fill", "parameters", "seed"),
        [
            ("Rotate", 90, 1.0, [0, 0, 0], {"limit": 90, "p": 1.0, "fill": [0, 0, 0]}, 1),
        ],
    )
    def test_configuration(
        self,
        perturber: str,
        limit: float | tuple[float, float],
        probability: float,
        fill: np.ndarray,
        parameters: dict[str, Any],
        seed: int,
    ) -> None:
        """Test configuration stability."""
        inst = RandomRotationPerturber(
            limit=limit,
            probability=probability,
            seed=seed,
        )
        for i in configuration_test_helper(inst):
            assert i.perturber == perturber
            assert i.limit == limit
            assert i.probability == probability
            assert i.fill == fill
            assert i.parameters == parameters
            assert i.seed == seed


@mock.patch.object(RandomRotationPerturber, "is_usable")
def test_missing_deps(mock_is_usable: MagicMock) -> None:
    """Test that an exception is raised when required dependencies are not installed."""
    mock_is_usable.return_value = False
    assert not RandomRotationPerturber.is_usable()
    with pytest.raises(AlbumentationsImportError):
        RandomRotationPerturber()
