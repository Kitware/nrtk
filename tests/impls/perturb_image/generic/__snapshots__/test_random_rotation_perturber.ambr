# serializer version: 1
# name: TestRandomRotationPerturber.test_bbox_transform
  list([
    tuple(
      <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[27.9999999  26.99999996] max_vertex=[28.99999995 28.99999995]>,
      dict({
        'label': 1.0,
      }),
    ),
    tuple(
      <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[23.99999991 21.99999958] max_vertex=[26.99999996 27.9999999 ]>,
      dict({
        'label': 2.0,
      }),
    ),
  ])
# ---
